sound/soc/sof/intel/hda.c

```diff
@@ -1359,6 +1362,52 @@ struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
 			}
 		}
 
+		codec_type = snd_soc_acpi_intel_detect_amp_type(sdev->dev);
+
+		if (tplg_fixup &&
+		    mach->tplg_quirk_mask & SND_SOC_ACPI_TPLG_INTEL_AMP_NAME &&
+		    codec_type != CODEC_NONE) {
+			tplg_suffix = snd_soc_acpi_intel_get_amp_tplg_suffix(codec_type);
+			if (!tplg_suffix) {
+				dev_err(sdev->dev, "error: no tplg suffix found, amp %d\n",
+					codec_type);
+				return NULL;
+			}
+
+			tplg_filename = devm_kasprintf(sdev->dev, GFP_KERNEL,
+						       "%s-%s",
+						       sof_pdata->tplg_filename,
+						       tplg_suffix);
```

Comment:
Humm, I am a bit worried that we are adding two more branches to change the topology filename. In the end there will be 3 suffixes, are they all independent?

I think the initial one  SND_SOC_ACPI_TPLG_INTEL_SSP_NUMBER can only be used for the codec, it's a single-SSP configuration.

Reply:
Yes, they are independent and currently SND_SOC_ACPI_TPLG_INTEL_SSP_NUMBER flag is used in es8336 driver enumeration only.

Other machine drivers (mostly Chromebooks) could use 'tplg_filename' module parameter to overwrite the topology file name if they need to use something like '-sspX' instead of using SND_SOC_ACPI_TPLG_INTEL_SSP_NUMBER flag.

Reply:
> Yes, they are independent and currently SND_SOC_ACPI_TPLG_INTEL_SSP_NUMBER flag is used in es8336 driver enumeration only.

The reply above tells me all flags cannot be used concurrently, so they are not independent flags with orthogonal logic.

Reply:
Sorry for my misconception. I thought the output suffix string of a flag does not depend on the code branch of other flags so they are independent.

Do you foresee a situation that we have to use the three flags concurrently?

Reply:
No, like you said the SSP_NUMBER quirks was only meant to be used for the ES8336. Just adding a test to check that SSP_NUMBER is ignored if one of the others is set should work and avoid broken configurations leading to 'no topology file found'.

The point is that this needs to be user-proof with defensive programming. Some folks will set the quirk as -1 just because they can...

Reply:
Ok. Let me test following code segment and upload for review tomorrow.

diff --git a/sound/soc/sof/intel/hda.c b/sound/soc/sof/intel/hda.c
index 67a4d50cb1d7..a490f15d46be 100644
--- a/sound/soc/sof/intel/hda.c
+++ b/sound/soc/sof/intel/hda.c
@@ -1261,6 +1261,27 @@ void hda_set_mach_params(struct snd_soc_acpi_mach *mach,
        mach_params->dai_drivers = desc->ops->drv;
 }

+static int check_tplg_quirk_mask(struct snd_soc_acpi_mach *mach)
+{
+       u32 dmic_ssp_quirk;
+       u32 codec_amp_name_quirk;
+
+       /*
+        * In current implementation dmic and ssp quirks are designed for es8336
+        * machine driver and could not be mixed with codec name and amp name
+        * quirks.
+        */
+       dmic_ssp_quirk = mach->tplg_quirk_mask &
+                        (SND_SOC_ACPI_TPLG_INTEL_DMIC_NUMBER | SND_SOC_ACPI_TPLG_INTEL_SSP_NUMBER);
+       codec_amp_name_quirk = mach->tplg_quirk_mask &
+                        (SND_SOC_ACPI_TPLG_INTEL_AMP_NAME | SND_SOC_ACPI_TPLG_INTEL_CODEC_NAME);
+
+       if (dmic_ssp_quirk && codec_amp_name_quirk)
+               return -EINVAL;
+
+       return 0;
+}
+
 struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
 {
        u32 interface_mask = hda_get_interface_mask(sdev);
@@ -1288,6 +1309,17 @@ struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
                        tplg_fixup = true;
                }

+               /*
+                * Checking quirk mask integrity; some quirk flags could not be
+                * set concurrently.
+                */
+               if (tplg_fixup &&
+                   check_tplg_quirk_mask(mach)) {
+                       dev_err(sdev->dev, "Invalid tplg quirk mask 0x%x\n",
+                               mach->tplg_quirk_mask);
+                       return NULL;
+               }
+
                /* report to machine driver if any DMICs are found */
                mach->mach_params.dmic_num = check_dmic_num(sdev);